pkg/task/fetch.go

```diff
@@ -60,27 +78,124 @@ func NewK8sTaskSpecFetcher(searchNamespace string) (*K8sTaskSpecFetcher, error)
 	}, nil
 }
 
-// FetchSpec returns specifications of a provided task in yaml
-// string format
-//
-// NOTE:
-//  This is an implementation of TaskSpecFetcher interface
-func (f *K8sTaskSpecFetcher) Fetch(taskName string) (runtask v1alpha1.RunTask, err error) {
-	if len(taskName) == 0 {
-		err = fmt.Errorf("failed to fetch runtask: nil task name was provided")
+// runTaskGetter abstracts fetching of runtask instance
+type runTaskGetter interface {
+	Get() (runtask *v1alpha1.RunTask, err error)
+}
+
+// getRunTaskSpec composes common properties required to get a run task
+// instance
+type getRunTaskSpec struct {
+	taskName  string
+	k8sClient *m_k8s_client.K8sClient
+}
+
+// runTaskGetterFn abstracts fetching of runtask instance based on the provided
+// runtask specifications
+type runTaskGetterFn func(getSpec getRunTaskSpec) (runtask *v1alpha1.RunTask, err error)
+
+// getRunTaskFromConfigMap fetches runtask instance from a config map instance
+func getRunTaskFromConfigMap(g getRunTaskSpec) (runtask *v1alpha1.RunTask, err error) {
+	if len(strings.TrimSpace(g.taskName)) == 0 {
+		err = fmt.Errorf("missing run task name: failed to get runtask from config map")
+		return
+	}
+
+	if g.k8sClient == nil {
+		err = fmt.Errorf("nil kubernetes client found: failed to get runtask '%s' from config map", g.taskName)
 		return
 	}
 
-	cm, err := f.k8sClient.GetConfigMap(taskName, mach_apis_meta_v1.GetOptions{})
+	cm, err := g.k8sClient.GetConfigMap(g.taskName, mach_apis_meta_v1.GetOptions{})
 	if err != nil {
+		err = errors.Wrap(err, fmt.Sprintf("failed to get run task '%s' from config map", g.taskName))
 		return
 	}
 
-	runtask = v1alpha1.RunTask{}
-	runtask.Name = taskName
+	runtask = &v1alpha1.RunTask{}
+	runtask.Name = g.taskName
 	runtask.Spec.Meta = cm.Data["meta"]
 	runtask.Spec.Task = cm.Data["task"]
 	runtask.Spec.PostRun = cm.Data["post"]
 
 	return
 }
+
+// getRunTaskFromCustomResource fetches runtask instance from its custom
+// resource instance
+func getRunTaskFromCustomResource(g getRunTaskSpec) (runtask *v1alpha1.RunTask, err error) {
+	if len(strings.TrimSpace(g.taskName)) == 0 {
+		err = fmt.Errorf("missing run task name: failed to get runtask from custom resource")
+		return
+	}
+
+	if g.k8sClient == nil {
+		err = fmt.Errorf("nil kubernetes client: failed to get runtask '%s' from custom resource", g.taskName)
+		return
+	}
+
+	runtask, err = g.k8sClient.GetOEV1alpha1RunTask(g.taskName, mach_apis_meta_v1.GetOptions{})
+	if err != nil {
+		err = errors.Wrap(err, fmt.Sprintf("failed to get runtask '%s' from custom resource", g.taskName))
+	}
+
+	return
+}
+
+// getRunTask enables fetching a run task instance based on various run task
+// getter strategies
+//
+// NOTE:
+//  This is an implementation of runTaskGetter
+type getRunTask struct {
```

Comment:
I think we should consider some alternatives here. `getRunTask` sounds like a func

Reply:
Yes. It is. However this is exactly it is named. Single purpose object. I feel as long as compilation is handled it is good.